### PR TITLE
Query cache

### PR DESCRIPTION
**Pull request by mstetsyuk**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Query Cache. For more detailed information refer to https://github.com/ClickHouse/ClickHouse/issues/34011

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by kitaisreal** [comment]
@kssenii I am coordinator of this task.

**Comment by mo-avatar** [comment]
Looking forward to having this feature， is there an schedule for putting it into  experimental feature?

**Comment by kitaisreal** [comment]
@mo-avatar it will be production ready. Estimated time is 3-4 months. Details are in this ticket #34011.

**Comment by mo-avatar** [comment]
> @mo-avatar it will be production ready. Estimated time is 3-4 months. Details are in this ticket #34011.
Glad to hear. Thanks for your answer.

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 1]
```diff
@@ -102,6 +105,8 @@ namespace ErrorCodes
     extern const int ACCESS_DENIED;
 }
 
+std::unordered_map<IAST::Hash, Data, ASTHash> InterpreterSelectQuery::cached_data;
```
Comment (kitaisreal) [inline src/Interpreters/InterpreterSelectQuery.cpp:108]: std::unordered_map is slow, and also we have predefined cache, check `UncompressedCache.h` and try to reuse it.

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 2]
```diff
@@ -589,8 +594,23 @@ InterpreterSelectQuery::InterpreterSelectQuery(
 
 void InterpreterSelectQuery::buildQueryPlan(QueryPlan & query_plan)
 {
+    if (cached_data.contains(query_ptr->getTreeHash()))
```
Comment (kitaisreal) [inline src/Interpreters/InterpreterSelectQuery.cpp:597]: `query_ptr->getTreeHash()` need to be extracted in separate variable, because this method call is heavy.

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 3]
```diff
@@ -589,8 +594,23 @@ InterpreterSelectQuery::InterpreterSelectQuery(
...
+
+        Pipe pipe(std::make_shared<ReadFromCacheTransform>(header, chunks));
+        auto read_from_cache_step = std::make_unique<ReadFromPreparedSource>(std::move(pipe));
+        read_from_cache_step->setStepDescription("Read query result from cache");
+        query_plan.addStep(std::move(read_from_cache_step));
+        return;
+    }
     executeImpl(query_plan, std::move(input_pipe));
 
+    auto caching_step = std::make_unique<CachingStep>(query_plan.getCurrentDataStream(), cached_data, query_ptr);
```
Comment (kitaisreal) [inline src/Interpreters/InterpreterSelectQuery.cpp:610]: Currently better to only use cache if some setting is specified, for now we can make simple `use_query_cache`, check `Settings.h`.

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 4]
```diff
@@ -2619,5 +2639,20 @@ void InterpreterSelectQuery::initSettings()
 
     }
 }
+Chunk InterpreterSelectQuery::create_single_cache_chunk_from_many(IAST::Hash ASTHash)
```
Comment (kitaisreal) [inline src/Interpreters/InterpreterSelectQuery.cpp:2642]: This method will not work, chunk has multiple columns, and when you merge chunks you need to merge each column independently. Assume we have 2 chunks with 2 columns, then during merge we will have 1 chunk with 2 columns and each column will contain data from 2 chunks. Check `IColumn.h` and `insertFrom` methods.

File: src/Interpreters/InterpreterSelectQuery.h [hunk 5]
```diff
@@ -29,6 +32,15 @@ class QueryPlan;
 
 struct TreeRewriterResult;
 using TreeRewriterResultPtr = std::shared_ptr<const TreeRewriterResult>;
+using Data = std::pair<Block, Chunks>;
+
+struct ASTHash
+{
+    size_t operator() (const IAST::Hash & hash) const
+    {
+        return hash.first * 7919 + hash.second;
```
Comment (kitaisreal) [inline src/Interpreters/InterpreterSelectQuery.h:41]: In cache we should put not only AST but also header type, because we will get logical errors.
Consider example
```sql
CREATE TABLE test_table (id UInt64) ENGINE=TinyLog;
SELECT id FROM test_table;
DROP TABLE test_table;
CREATE TABLE test_table (id String) ENGINE=TinyLog;
SELECT id FROM test_table;
```
this will lead to crash.

File: tests/performance/file_table_function.xml [hunk 6]
```diff
@@ -0,0 +1,54 @@
+<test>
```
Comment (kitaisreal) [inline tests/performance/file_table_function.xml:1]: Remove changes to this file.

File: src/Processors/Transforms/CachingTransform.cpp [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+{
+
+void CachingTransform::transform(Chunk & chunk)
+{
+    if (!fits_into_memory)
+    {
+        return;
+    }
+
+    fits_into_memory = cache->insertChunk(cache_key, chunk.clone());
```
Comment (kitaisreal) [inline src/Processors/Transforms/CachingTransform.cpp:13]: Recalculate this `fits_into_memory`, using settings from Settings.h

Comment (kitaisreal) [reply src/Processors/Transforms/CachingTransform.cpp:13]: Need to be prepare if some exception will be thrown, lets that also be handled as part of API.

File: src/Storages/QueryCache.h [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+struct CacheKeyHasher
+{
+    size_t operator()(const CacheKey & key) const
+    {
+        auto ast_info = key.ast->getTreeHash();
+        auto header_info = std::hash<String>{}(key.header.getNamesAndTypesList().toString());
+        auto settings_info = settingsHash(key.settings);
+        auto username_info = std::hash<std::optional<String>>{}(key.username);
+
+        return ast_info.first + ast_info.second * 9273
```
Comment (kitaisreal) [inline src/Storages/QueryCache.h:45]: Consider to use SipHash, our boost hash combine. Do not make any manual logic with hashing.

File: src/Storages/QueryCache.h [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+    private:
+        static size_t settingsHash(const Settings & settings)
+        {
+            size_t hash = 0;
+            size_t coefficient = 1;
+            for (const auto & setting : settings)
+            {
+                hash += std::hash<String>{}(setting.getValueString()) * coefficient;
+                coefficient *= 53;
```
Comment (kitaisreal) [inline src/Storages/QueryCache.h:59]: Do not try to hash settings manually.

File: src/Storages/QueryCache.h [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+            }
+            return hash;
+        }
+};
+
+struct QueryWeightFunction
+{
+    size_t operator()(const Data & data) const
```
Comment (kitaisreal) [inline src/Storages/QueryCache.h:67]: Weight of data should be member of data. Because this QueryWeightFunction can be called a lot of times in LRUCache during evictions.

File: src/Storages/QueryCache.h [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+private:
+    using timestamp = std::chrono::time_point<std::chrono::high_resolution_clock>;
+    using duration = std::chrono::high_resolution_clock::duration;
+public:
+    void scheduleRemoval(duration duration, CacheKey cache_key)
+    {
+        std::lock_guard lock(mutex);
+        TimedCacheKey timer = {now() + duration, cache_key};
+        queue.push(timer);
+        if (queue.top() == timer)
```
Comment (kitaisreal) [inline src/Storages/QueryCache.h:94]: In queue we need to store timers in increasing order of timestamps. We should notify thread only if our new sheduled value is less than current minimum.
Please add comment for this. Also notifying of condition variable can be done not under mutex here.

File: src/Storages/QueryCache.h [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+        {
+            timer_cv.notify_one();
+        }
+    }
+
+    template <typename Cache>
+    [[noreturn]] void processRemovalQueue(Cache * query_cache)
```
Comment (kitaisreal) [inline src/Storages/QueryCache.h:101]: Remove [[noreturn]] add special is_finished atomic flag. This thread need to be joined manually during destruction.

File: src/Storages/QueryCache.h [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+        res += block.allocatedBytes();
+
+        return res;
+    }
+};
+
+class CacheRemovalScheduler
+{
+private:
+    using timestamp = std::chrono::time_point<std::chrono::high_resolution_clock>;
```
Comment (kitaisreal) [inline src/Storages/QueryCache.h:86]: using `Timestamp`.

File: src/Storages/QueryCache.h [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+    using duration = std::chrono::high_resolution_clock::duration;
```
Comment (kitaisreal) [inline src/Storages/QueryCache.h:87]: using `Duration`.

File: src/Storages/QueryCache.h [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+        }
+    }
+
+    template <typename Cache>
+    [[noreturn]] void processRemovalQueue(Cache * query_cache)
+    {
+        while (true)
+        {
+            std::unique_lock lock(mutex);
+            const std::optional<TimedCacheKey> awaited_timer = nextTimer();
```
Comment (kitaisreal) [inline src/Storages/QueryCache.h:106]: This place is very hard to read to verify that there is no bugs. Please add comments and simplify logic.

File: src/Storages/QueryCache.h [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+
+            timer_cv.wait_until(lock,
+                                awaited_timer.has_value() ? awaited_timer->time : infinite_time,
+                                [&]() { return awaited_timer != nextTimer() || (awaited_timer.has_value() && awaited_timer->time <= now()); }
+            );
+
+            if (awaited_timer.has_value() && awaited_timer->time <= now())
+            {
+                query_cache->remove(awaited_timer->cache_key);
```
Comment (kitaisreal) [inline src/Storages/QueryCache.h:115]: This should be done without mutex lock. Because we block threads who try to schedule removal.

File: src/Storages/QueryCache.h [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+        return std::chrono::high_resolution_clock::now();
+    }
+
+    const timestamp infinite_time = timestamp::max();
+    std::priority_queue<TimedCacheKey> queue;
+    std::condition_variable timer_cv;
+    std::mutex mutex;
+};
+
+class QueryCache : public LRUCache<CacheKey, Data, CacheKeyHasher, QueryWeightFunction>
```
Comment (kitaisreal) [inline src/Storages/QueryCache.h:164]: Lets try to hide all implementation details inside .cpp file

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: programs/server/Server.cpp [hunk 18]
```diff
@@ -1373,6 +1373,11 @@ int Server::main(const std::vector<std::string> & /*args*/)
     if (index_mark_cache_size)
         global_context->setIndexMarkCache(index_mark_cache_size);
 
+    size_t query_cache_size_in_bytes = config().getUInt64("query_cache_size_in_bytes", /*1ULL << 30*/ 16000);
+    size_t max_query_cache_entry_size = config().getUInt64("max_entry_size_in_bytes", /*1ULL << 30*/ 14000);
```
[the remark this answers is not in the document]
Comment (mstetsyuk) [reply programs/server/Server.cpp]: Done. Moved the setting to Settings.h.

File: src/Common/LRUCache.h [hunk 19]
```diff
@@ -257,7 +257,6 @@ class LRUCache
 
     friend struct InsertTokenHolder;
 
-
```
[the remark this answers is not in the document]
Comment (mstetsyuk) [reply src/Common/LRUCache.h]: Done.

[the remark this answers is not in the document]
Comment (mstetsyuk) [reply src/Common/LRUCache.h]: Done.

File: src/Storages/QueryCache.h (hunk 8 above)
Comment (mstetsyuk) [reply src/Storages/QueryCache.h:45]: Done.

File: src/Storages/QueryCache.h (hunk 11 above)
Comment (mstetsyuk) [reply src/Storages/QueryCache.h]: Added a couple of comments.

File: src/Storages/QueryCache.h (hunk 14 above)
Comment (mstetsyuk) [reply src/Storages/QueryCache.h]: Fixed this.

File: src/Storages/QueryCache.h (hunk 16 above)
Comment (mstetsyuk) [reply src/Storages/QueryCache.h]: Fixed this.

File: src/Storages/QueryCache.h (hunk 12 above)
Comment (mstetsyuk) [reply src/Storages/QueryCache.h]: Done.

File: src/Storages/QueryCache.h [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+                lock.unlock();
+                queue.pop();
+                query_cache->remove(awaited_timer->cache_key);
+            }
+        }
+    }
+
+    void stopProcessingRemovalQueue()
+    {
+        process_removal_queue.store(false);
```
Comment (kitaisreal) [inline src/Storages/QueryCache.h:119]: It should be done protected by mutex.

Comment (kitaisreal) [inline src/Storages/QueryCache.h:119]: Better to rename flag into `is_finished`.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 21]
```diff
@@ -2775,5 +2816,23 @@ void InterpreterSelectQuery::initSettings()
 
     }
 }
+Chunk InterpreterSelectQuery::to_single_chunk(const Chunks& chunks)
```
[the remark this answers is not in the document]
Comment (mstetsyuk) [reply src/Interpreters/InterpreterSelectQuery.cpp]: Removed it completely from this class. It is now in CacheReadHolder.

File: src/Interpreters/InterpreterSelectQuery.cpp (hunk 2 above)
Comment (mstetsyuk) [reply src/Interpreters/InterpreterSelectQuery.cpp]: Done.

File: src/Interpreters/InterpreterSelectQuery.cpp (hunk 3 above)
Comment (mstetsyuk) [reply src/Interpreters/InterpreterSelectQuery.cpp]: Added two options for that: query_cache_active_usage and query_cache_passive_usage.

File: src/Interpreters/InterpreterSelectQuery.cpp (hunk 4 above)
Comment (mstetsyuk) [reply src/Interpreters/InterpreterSelectQuery.cpp]: Fixed.

File: src/Interpreters/InterpreterSelectQuery.h (hunk 5 above)
Comment (mstetsyuk) [reply src/Interpreters/InterpreterSelectQuery.h]: Fixed.

File: tests/performance/file_table_function.xml (hunk 6 above)
Comment (mstetsyuk) [reply tests/performance/file_table_function.xml]: Done.

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 22]
```diff
@@ -620,10 +623,48 @@ InterpreterSelectQuery::InterpreterSelectQuery(
...
+
+    auto caching_step = std::make_unique<CachingStep>(query_plan.getCurrentDataStream(),
+                                                      context->getQueryCache(), query_cache_key);
+    caching_step->setStepDescription("Put query result in cache");
+    query_plan.addStep(std::move(caching_step));
+}
+
 void InterpreterSelectQuery::buildQueryPlan(QueryPlan & query_plan)
 {
+    auto query_cache_key = CacheKey{query_ptr, source_header, context->getSettingsRef(),
```
[the remark this answers is not in the document]
Comment (mstetsyuk) [reply src/Interpreters/InterpreterSelectQuery.cpp]: Created two functions: executeReadFromCache and executePutInCache. CacheKey object is constructed inside the functions.

File: src/Processors/QueryPlan/CachingStep.cpp [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+            },
+            {
+                .preserves_number_of_rows = true,
+            }
+        };
+}
+
+void CachingStep::transformPipeline(QueryPipelineBuilder & pipeline, const BuildQueryPipelineSettings &)
+{
+    if (cache->getPutInCacheMutex(cache_key).try_lock())
```
[the remark this answers is not in the document]
Comment (mstetsyuk) [reply src/Processors/QueryPlan/CachingStep.cpp]: Turned out we don't actually need this mutex at all.

File: src/Processors/QueryPlan/CachingStep.cpp [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+            [&](const Block & header, QueryPipelineBuilder::StreamType) -> ProcessorPtr
+            { return std::make_shared<CachingTransform>(header, cache, cache_key.ast, cache_key.settings, cache_key.username); });
+    }
+}
+
+CachingStep::~CachingStep()
+{
+    if (execute_caching)
+    {
+        cache->getPutInCacheMutex(cache_key).unlock();
```
[the remark this answers is not in the document]
Comment (mstetsyuk) [reply src/Processors/QueryPlan/CachingStep.cpp]: The mutex is no longer used, so there's no manual lock-unlock anymore.

scheduleRemoval gets called in CachePutHandler's dtor.

File: src/Storages/QueryCache.h (hunk 11 above)
Comment (mstetsyuk) [reply src/Storages/QueryCache.h]: Added a plenty of comments to processRemovalQueue.

Comment (mstetsyuk) [reply src/Storages/QueryCache.h]: Added lock.unlock() before notifying condition variable.

File: src/Storages/QueryCache.h (hunk 20 above)
Comment (mstetsyuk) [reply src/Storages/QueryCache.h:130]: Done.

**Comment by alexey-milovidov** [comment]
@UnamedRus shared his thoughts about this feature with me in Telegram chat.
He wants this feature to be merged sooner and also asks - if subquery caching will be available.

**Comment by mo-avatar** [comment]
Hi guys， would you mind to tell if we can still expect the pr be merged recently (say one or two  months),  It seems that quite amount of  CI checks are still in failiing status for now.  In the mean time, great thanks for all your efforts, and nice work! @mstetsyuk @kitaisreal

**Comment by alexey-milovidov** [comment]
This PR is one of the most highly anticipated. It has a chance to become one of the best ClickHouse features in the second half of 2022. I want it to be merged and will be very happy to make it sooner!

**Comment by rschu1ze** [comment]
Will try to take a look in the coming days

File: src/Storages/QueryCache.h [hunk 25]
```diff
@@ -0,0 +1,351 @@
...
+    void clearRemovalQueue()
+    {
+        std::lock_guard lock(mutex);
+        queue = {};
+        timer_cv.notify_one();
+    }
+
+
+private:
+    struct TimedCacheKey
```
Comment (rschu1ze) [inline src/Storages/QueryCache.h:144]: `CacheKeyWithTimestamp`?

File: src/Storages/QueryCache.h (hunk 15 above)
Comment (rschu1ze) [reply src/Storages/QueryCache.h:114]: The logic is still too complex.

If I understand correctly, the purpose of CacheRemovalScheduler is to periodically remove old entries from the cache. Why can't they just sit there? The cache's LRU eviction policy will automatically remove the oldest entries.

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 26]
```diff
@@ -718,10 +720,64 @@ InterpreterSelectQuery::InterpreterSelectQuery(
     sanitizeBlock(result_header, true);
 }
 
+void InterpreterSelectQuery::executePutInCache(QueryPlan & query_plan)
+{
+    auto settings = context->getSettingsRef();
+    QueryCachePtr query_cache = context->getQueryCache();
+    if (!settings.query_cache_active_usage || query_cache == nullptr)
```
Comment (rschu1ze) [inline src/Interpreters/InterpreterSelectQuery.cpp:727]: I guess if the cache is currently `nullptr` (because the database started with `query_cache_active_usage = false`)  and the user just configured `query_cache_active_usage = true`, then we should create the cache and not silently return.

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 27]
```diff
@@ -1298,7 +1354,7 @@ void InterpreterSelectQuery::executeImpl(QueryPlan & query_plan, std::optional<P
     }
 
     if (query_info.projection && query_info.projection->input_order_info && query_info.input_order_info)
-       throw Exception("InputOrderInfo is set for projection and for query", ErrorCodes::LOGICAL_ERROR);
+        throw Exception("InputOrderInfo is set for projection and for query", ErrorCodes::LOGICAL_ERROR);
```
Comment (rschu1ze) [inline src/Interpreters/InterpreterSelectQuery.cpp:1357]: Let's not change unrelated code.

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 28]
```diff
@@ -718,10 +720,64 @@ InterpreterSelectQuery::InterpreterSelectQuery(
...
+    {
+        return;
+    }
+
+    auto caching_step = std::make_unique<CachingStep>(query_plan.getCurrentDataStream(), context->getQueryCache(), query_cache_key);
+    caching_step->setStepDescription("Put query result in cache");
+    query_plan.addStep(std::move(caching_step));
+}
+
+bool InterpreterSelectQuery::executeReadFromCache(QueryPlan & query_plan)
```
Comment (rschu1ze) [inline src/Interpreters/InterpreterSelectQuery.cpp:748]: What happens if the underlying data changed? That should invalidate the result cache, right?

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 29]
```diff
@@ -718,10 +720,64 @@ InterpreterSelectQuery::InterpreterSelectQuery(
...
+void InterpreterSelectQuery::executePutInCache(QueryPlan & query_plan)
+{
+    auto settings = context->getSettingsRef();
+    QueryCachePtr query_cache = context->getQueryCache();
+    if (!settings.query_cache_active_usage || query_cache == nullptr)
+    {
+        return;
+    }
+
+    auto query_cache_key = CacheKey{query_ptr, result_header, settings,
```
Comment (rschu1ze) [inline src/Interpreters/InterpreterSelectQuery.cpp:732]: The MySQL query result cache (which they killed in the meantime) excluded queries with non-deterministic functions (https://dev.mysql.com/doc/refman/5.6/en/query-cache-operation.html). I think we should do that as well (and additionally I'd also put all encryption/decryption functions on the blocklist so that we never store cleartext in the cache - this could get funny with user sharing).

**Comment by rschu1ze** [comment]
I'll continue this PR [here](https://github.com/ClickHouse/ClickHouse/pull/43797).
